File: pythonlibrary/aibrd/analyzers/gap_detector.py
from datetime import date
from ..llm.client import call_llm_json
from ..models.outputs import GapResult

_SYSTEM = """You are a software auditor comparing requirements against code.
For each requirement, determine if the code covers it.
Return JSON: {"gaps": [{"requirement_id": "...", "requirement_summary": "...",
"status": "covered|partial|missing", "reason": "..."}]}"""


def detect_gaps(context_content: str, code_snippets: str) -> list[GapResult]:
    max_chars = 12000
    prompt = (
        f"REQUIREMENTS:\n{context_content[:max_chars // 2]}\n\n"
        f"CODE:\n{code_snippets[:max_chars // 2]}"
    )
    raw = call_llm_json(prompt, _SYSTEM)
    return [GapResult(**g) for g in raw.get("gaps", [])]


def format_gap_report(gaps: list[GapResult]) -> str:
    today = date.today().isoformat()
    missing = [g for g in gaps if g.status == "missing"]
    partial = [g for g in gaps if g.status == "partial"]
    covered = [g for g in gaps if g.status == "covered"]

    lines = [
        "# Gap Report",
        f"_Generated: {today}_",
        "",
        f"**Summary:** {len(covered)} covered · {len(partial)} partial · {len(missing)} missing",
        "",
    ]
    if missing:
        lines += ["## ❌ Missing Coverage", ""]
        for g in missing:
            lines += [f"### {g.requirement_id}: {g.requirement_summary}", g.reason, ""]
    if partial:
        lines += ["## ⚠️ Partial Coverage", ""]
        for g in partial:
            lines += [f"### {g.requirement_id}: {g.requirement_summary}", g.reason, ""]

    return "\n".join(lines)

File: vscodebase/src/core/analyzers/gap_detector.ts
import * as vscode from 'vscode'
import { callLLMJson } from '../../llm/client'
import { buildGapPrompt } from '../../llm/context_builder'

export interface GapResult {
  requirementId: string
  requirementSummary: string
  status: 'covered' | 'partial' | 'missing'
  reason: string
}

const SYSTEM = `You are a software auditor. Compare requirements against code.
For each requirement, determine if the code covers it.
Return JSON: {
  "gaps": [{
    "requirementId": "...",
    "requirementSummary": "...",
    "status": "covered|partial|missing",
    "reason": "..."
  }]
}`

export async function detectGaps(
  contextContent: string,
  codeSnippets: string,
  token?: vscode.CancellationToken
): Promise<GapResult[]> {
  const prompt = buildGapPrompt(contextContent, codeSnippets)
  const result = await callLLMJson<{ gaps: GapResult[] }>(prompt, SYSTEM, token)
  return result.gaps
}

export function formatGapReport(gaps: GapResult[]): string {
  const lines: string[] = [
    '# Gap Report',
    `_Generated: ${new Date().toISOString().split('T')[0]}_`,
    ''
  ]

  const missing = gaps.filter(g => g.status === 'missing')
  const partial = gaps.filter(g => g.status === 'partial')
  const covered = gaps.filter(g => g.status === 'covered')

  lines.push(`**Summary:** ${covered.length} covered · ${partial.length} partial · ${missing.length} missing`, '')

  if (missing.length > 0) {
    lines.push('## ❌ Missing Coverage', '')
    for (const g of missing) {
      lines.push(`### ${g.requirementId}: ${g.requirementSummary}`)
      lines.push(`${g.reason}`, '')
    }
  }

  if (partial.length > 0) {
    lines.push('## ⚠️ Partial Coverage', '')
    for (const g of partial) {
      lines.push(`### ${g.requirementId}: ${g.requirementSummary}`)
      lines.push(`${g.reason}`, '')
    }
  }

  return lines.join('\n')
}
